internal/pkg/cli/task_run.go

```diff
@@ -134,6 +176,155 @@ func (o *runTaskOpts) Ask() error {
 	return nil
 }
 
+// Execute create or update task resources and run the task.
+func (o *runTaskOpts) Execute() error {
+	if o.image == "" && o.dockerfilePath == "" {
+		o.dockerfilePath = defaultDockerfilePath
+	}
+
+	if o.imageTag == "" {
+		o.imageTag = defaultImageTag
+	}
+
+	if err := o.getNetworkConfig(); err != nil {
+		return err
+	}
+
+	cluster, err := o.getCluster()
+	if err != nil {
+		return err
+	}
+
+	if err := o.createAndUpdateTaskResources(); err != nil {
+		return err
+	}
+
+	//if image is not provided, then we build the image and push to ECR repo
+	if o.image == "" {
+		uri, err := o.pushToECRRepo()
+		if err != nil {
+			return err
+		}
+		o.image = fmt.Sprintf(fmtImageURL, uri, o.imageTag)
+
+		// update image to stack
+		if err := o.createAndUpdateTaskResources(); err != nil {
+			return err
+		}
+	}
+
+	return o.runTask(cluster)
+}
+
+func (o *runTaskOpts) getCluster() (string, error) {
+	if o.env == config.EnvNameNone {
+		clusters, err := o.ecsGetter.DefaultClusters()
+		if err != nil {
+			return "", fmt.Errorf("get default cluster: %w", err)
+		}
+		return clusters[0], nil
+	}
+
+	clusters, err := o.resourceGetter.GetResourcesByTags(clusterResourceType, map[string]string{
+		deploy.AppTagKey: o.AppName(),
+		deploy.EnvTagKey: o.env,
+	})
+
+	if err != nil {
+		return "", fmt.Errorf("get cluster: %w", err)
+	}
+
+	return clusters[0], nil
```

Comment:
what if the length of clusters is 0? it is possible if `GetResourcesByTags` cannot get any matching result.

Reply:
if length is 0, `DefaultClusters()` in `ecs` package would have returned an error, which would have been passed through `getCluster()`, so this line wouldn't have been reached

Reply:
I mean `clusters` defined here https://github.com/aws/copilot-cli/pull/1077/files#diff-d83adc990a0da087a39c02967f267ed3R228. Not the one for default clusters.